src/intervals/exact_numbers.jl

```diff
@@ -0,0 +1,68 @@
+struct ExactNumber{T} <: Real
+    number::T
+end
+
+# Promotion to Interval
+Base.promote_rule(::Type{Interval{T}}, ::Type{ExactNumber{S}}) where {T<:NumTypes,S<:Real} =
+    Interval{promote_numtype(T, S)}
+
+Base.promote_rule(::Type{ExactNumber{T}}, ::Type{Interval{S}}) where {T<:Real,S<:NumTypes} =
+    Interval{promote_numtype(T, S)}
+
+function Base.convert(::Type{Interval{T}}, x::ExactNumber) where {T<:NumTypes}
+    y = interval(T, x)
+    return _unsafe_interval(bareinterval(y), decoration(y), true)
+end
+
+# Promotion to Real
+Base.promote_rule(::Type{T}, ::Type{ExactNumber{S}}) where {T<:Real,S<:Real} =
+    promote_type(T, S)
+
+Base.promote_rule(::Type{ExactNumber{T}}, ::Type{S}) where {T<:Real,S<:Real} =
+    promote_type(T, S)
+
+Base.convert(::Type{T}, x::ExactNumber) where {T<:Real} =
+    convert(T, x.number)
+
+interval(::Type{T}, x::ExactNumber) where {T<:NumTypes} = interval(T, x.number)
+interval(x::ExactNumber) = interval(x.number)
+
+Base.string(x::ExactNumber{T}) where {T<:AbstractFloat} =
+    Base.Ryu.writefixed(x.number, 2000, false, false, false, UInt8('.'), true)
+
+Base.string(x::ExactNumber) = string(x.number)
+
+Base.show(io::IO, ::MIME"text/plain", x::ExactNumber{T}) where {T<:AbstractFloat} =
+    print(io, "ExactNumber{$T}($(string(x)))")
+
+for unary in Symbol.((inv, sqrt, sin, cos, tan, asin, acos, atan, exp, log, sinh, cosh, tanh, asinh, acosh, atanh))
+    @eval (Base.$unary)(x::ExactNumber) = ($unary)(interval(x))
+end
+
+for binary in Symbol.((+, -, *, /, ^))
+    @eval (Base.$binary)(x::ExactNumber, y::ExactNumber) = ($binary)(interval(x), interval(y))
+end
+
+macro exact(expr)
+    exact_expr = postwalk(expr) do x
+        if x isa Real
+            e = ExactNumber(x)
+            s = string(e)
+            if s != string(x)
+                @warn "$(typeof(x)) displayed as $x is equal to $s"
```

Comment:
My first impulse would be to not have yet an other warning being prompted. If the user types in explicitly @exact, then it's on them that they know about floats.